src/day20.scala

```diff
@@ -0,0 +1,114 @@
+// using scala 3.1.0
+
+package day20
+
+import scala.util.Using
+import scala.io.Source
+import scala.annotation.tailrec
+
+@main def part1(): Unit =
+  println(s"The solution is ${part1(readInput())}")
+
+@main def part2(): Unit =
+  println(s"The solution is ${part2(readInput())}")
+
+def readInput(): String =
+  Using.resource(Source.fromFile("input/day20"))(_.mkString)
+
+def part1(input: String): Int =
+  val (enhancer, image0) = parseEnhancerAndImage(input)
+  val image1 = enhancer.enhance(image0)
+  val image2 = enhancer.enhance(image1)
+  image2.countLitPixels()
+
+def parseEnhancerAndImage(input: String): (Enhancer, Image) =
+  val enhancerAndImage = input.split("\n\n")
+  val enhancer         = Enhancer.parse(enhancerAndImage(0))
+  val image            = Image.parse(enhancerAndImage(1))
+  (enhancer, image)
+
+enum Pixel:
+  case Lit, Dark
+
+object Pixel:
+
+  def parse(char: Char): Pixel =
+    char match
+      case '#' => Pixel.Lit
+      case '.' => Pixel.Dark
+
+end Pixel
+
+object Enhancer:
+
+  def parse(input: String): Enhancer =
+    Enhancer(input.map(Pixel.parse))
+
+end Enhancer
+
+class Enhancer(enhancementString: IndexedSeq[Pixel]):
+
+  def enhance(image: Image): Image =
+    val pixels =
+      for y <- -1 until (image.height + 1)
+      yield
+        for x <- -1 until (image.width + 1)
+        yield enhancementString(pixelValue(image, x, y))
+
+    val outboundsPixel =
+      val value = if image.outboundsPixel == Pixel.Dark then 0 else 511
+      enhancementString(value)
+
+    Image(pixels, outboundsPixel)
+  end enhance
+
+  private def pixelValue(image: Image, x: Int, y: Int): Int =
+    var result  = 0
+    var bitMask = 1
+    for
+      yy <- (y + 1) to (y - 1) by -1
+      xx <- (x + 1) to (x - 1) by -1
+    do
+      if image.pixel(xx, yy) == Pixel.Lit then
+        result = result | bitMask
+      end if
+      bitMask = bitMask << 1
+    result
+  end pixelValue
```

Comment:
It took me a moment to realize why you were going `by -1`. It's because you're shifting the bit mask to the left on every iteration.

Consider instead shifting the *result* to the left on each iteration, as follows:
```suggestion
  private def pixelValue(image: Image, x: Int, y: Int): Int =
    var result = 0
    for
      yy <- (y + 1) to (y - 1)
      xx <- (x + 1) to (x - 1)
    do
      result = result << 1
      if image.pixel(xx, yy) == Pixel.Lit then
        result = result | 1
      end if
    result
  end pixelValue
```
That way, it's not necessary to maintain `bitMask` at all.

Reply:
Also, if you use `end if`, you should use `end for`, since the `for` encloses the `if`.

Reply:
Good point, thanks for the suggestion. Fixed in 4a9a2e7.